Documentation/design-docs/tailcalls-with-helpers.md

```diff
@@ -0,0 +1,375 @@
+```
+* Return address of A
+* Callee saved registers and locals of A
+* Stack arguments of B
+* Return address of B
+```
+### In CallTarget thunk for D, about to call D
+The thunk will now extract parameters for D from the thread local storage and call D.
+```
+* Return address of A
+* Callee saved registers and locals of A
+* Stack arguments of B
+* Return address of B
+* Callee saved registers and locals of CallTarget thunk for D
+```
+### In D
+We are in the last function of the chain, so after it does its work, it returns to its CallTarget thunk. 
+```
+* Return address of A
+* Callee saved registers and locals of A
+* Stack arguments of B
+* Return address of B
+* Callee saved registers and locals of CallTarget thunk for D
+* Stack arguments of D
+* Return address of D
+* Callee saved registers and locals of D
+```
+### Returned to CallTarget thunk for D with ChainCall flag in the TailCallHelperStackEntry **clear**
+The thunk checks the ChainCall flag and since it is clear, it recognizes we are now returning from the call chain and so it returns the result of the D. 
+```
+* Return address of A
+* Callee saved registers and locals of A
+* Stack arguments of B
+* Return address of B
+* Callee saved registers and locals of CallTarget thunk for D
+```
+### Returned to A
+We are back in A and we have the return value of the call chain.
+```
+* Return address of A
+* Callee saved registers and locals of A
+```
+
+## Example of thunks generated for a simple generic method
+
+```C#
+struct Point
+{
+    public int x;
+    public int y;
+    public int z;
+}
+
+class Foo
+{
+    public Point Test<T>(int x, T t) where T : class
+    {
+        Console.WriteLine("T: {0}", typeof(t));
+        return new Point(x, x, x);
+    }
+}
+```
+
+For tail calling the Test function, the IL helpers could look e.g. as follows:
+
+```IL
+.method public static Point StoreArgumentsTest(object thisObj, object genericContext, int x, object t) cil managed
+{
+	.maxstack 4
+	.locals init (
+		[0] valuetype Point,
+	)
+
+    call native int FetchArgumentBuffer()
+
+	dup
+	ldarg.0 // "thisObj"
+	stobj object
+	sizeof object
+	add
+	dup
+
+    ldarg.1 // "genericContext"
+	stobj object
+	sizeof object
+	add
+	dup
+
+	ldarg.2 // "x"
+	stind.i
+	sizeof int32
+	add
+	dup
+
+    ldarg.3 // "t"
+	stobj object
+
+	ldloca.s 0 // return default(Point)
+	initobj Point
+	ldloc.0
+
+	ret
+}
+
+```
+
+```IL
+.method public static Point CallTargetTest() cil managed
+{
+    .maxstack 4
+    .locals init (
+       [0] native int buffer
+    )
+    call native int FetchArgumentBuffer()
+    stloc.0
+
+    ldloc.0
+    ldobj object // this
+    ldloc.0
+    sizeof object
+    add
+    stloc.0
+
+    ldloc.0
+    ldobj object // generic context
+    ldloc.0
+    sizeof object
+    add
+    stloc.0
+
+    ldloc.0
+    ldind.i // int x
+    ldloc.0
+    sizeof int32
+    add
+    stloc.0
+
+    ldloc.0
+    ldobj object // T t
+
+    calli Point (object, object, int32, object)
```

Comment:
Ah, the address loading is missing in this example. I have added these examples as a last minute idea and now I can see I should have spent more time on them. 
The calli is there so that the call can be blind and we can pass this and generic context explicitly (I may be wrong, but I think the regular call doesn't allow this)